base/utftype.jl

```diff
@@ -0,0 +1,39 @@
+# This file is a part of Julia. License is MIT: http://julialang.org/license
+
+#=#!
+@brief      Base UTF16String type, has 16-bit NULL termination word after data, native byte order
+
+@throws     UnicodeError
+=#
+immutable UTF16String <: AbstractString
+    data::Vector{UInt16} # includes 16-bit NULL termination after string chars
+    function UTF16String(data::Vector{UInt16})
+        if length(data) < 1 || data[end] != 0
+            throw(UnicodeError(UTF_ERR_NULL_16_TERMINATE, 0, 0))
+        end
+        new(data)
+    end
+end
+
+#=#!
+@brief      Base UTF32String type, has 32-bit NULL termination word after data, native byte order
+
+@throws     UnicodeError
+=#
+immutable UTF32String <: DirectIndexString
+    data::Vector{Char} # includes 32-bit NULL termination after string chars
+
+    function UTF32String(data::Vector{Char})
+        if length(data) < 1 || data[end] != Char(0)
+            throw(UnicodeError(UTF_ERR_NULL_32_TERMINATE, 0, 0))
+        end
+        new(data)
+    end
+end
+UTF32String(data::Vector{UInt32}) = UTF32String(reinterpret(Char, data))
+
+const empty_utf16 = UTF16String(UInt16[0])
+const empty_utf32 = UTF32String(UInt32[0])
+
+isvalid{T<:Union(ASCIIString,UTF8String,UTF16String,UTF32String)}(str::T) = isvalid(T, str.data)
+isvalid{T<:Union(ASCIIString,UTF8String,UTF16String,UTF32String)}(::Type{T}, str::T) = isvalid(T, str.data)
```

Comment:
Let's add `empty_utf16` and `empty_utf32` in the same commit that first uses them, not here.

I found the definitions for `isvalid` here - https://github.com/ScottPJones/julia/blob/544801b43e943887e6fa16ae296211ecb51066d4/base/utf32.jl#L80-L81

You're right that `utftype[s].jl` is a better place for them. But they should be deleted from `utf32.jl`, not duplicated in 2 places.